### PR TITLE
Ported to MIPS CI20 Creator

### PR DESCRIPTION
**Pull request by garysims**
Initial port to MIPS CI20 Creator.

### CODE REVIEW

File: wiringX.c [hunk 1]
```diff
@@ -409,6 +410,37 @@ int wiringXSetup(void) {
 	if(wiringXLog == NULL) {
 		wiringXLog = _fprintf;
 	}
+
+#ifdef __mips__
```
Comment (CurlyMoo) [inline wiringX.c]: Can you merge this:
`#if defined(__arm__) || defined(__mips__)`

**Comment by CurlyMoo** [comment]
What devices does this support?

**Comment by garysims** [comment]
MIPS CI20 Creator:

http://www.elinux.org/MIPS_Creator_CI20

http://store.imgtec.com/us/product/mips-creator-ci20/

Gary

**Comment by CurlyMoo** [comment]
Do the example programs work as expected?

**Comment by garysims** [comment]
blink.c works, I haven't tried the other two.

**Comment by CurlyMoo** [comment]
Let me know when you did. If i remembered correctly the interrupt output must be:

```
interrupt
interrupt
timeout
interrupt
interrupt
timeout
```

etc.

**Comment by garysims** [comment]
Looking at the PIN diagram for the CI20 here: http://www.elinux.org/CI20_Hardware (in the Primary expansion header section) - What pin should wiringXISR() use for interrupt.c?

**Comment by CurlyMoo** [comment]
You should map the gpio numbers to sysfs numbers for that in the module. See this example:
http://www.elinux.org/CI20_GPIO_LED_Blink_Tutorial#Example:_blinking_LED

As you can see here:
https://github.com/wiringX/wiringX/blob/master/hummingboard.c#L50
GPIO 0 corresponds to sysfs 73

**Comment by garysims** [comment]
You miss understand me. I have the mapping code in ci20.c:

```
static int pinToGpio[27] = {
    -1, // Pin 0 - doesn't exist
    -1, -1, 126, -1, 127, -1, 124, 163,     // Pins 1 to 8
    -1, 160, 122, 133, 123, -1, 125, 161,   // Pins 9 to 16
    -1, 162, 145, -1, 142, 136, 143, 144,   // Pins 10 to 24
    -1, 146                                 // Pins 25 and 26
};
```

So pin 7 is GPIO1 / PD28 or sysfs 124. And the code works as the LED blinks as expected. However, there is no PIN 0 in the CI20, it starts at 1.

But my real question is this, I don't understand what interrupt.c is doing. What is coming in on PIN 1? Is it a clock? It is a link from PIN0 so that when the LED lights PIN1 is high? Is there a special circuit?

Unfortunately hardware isn't my strong point, my degree was in software engineering.

**Comment by garysims** [comment]
Also according to this: https://blog.pixxis.be/2014/08/09/humming_board_gpio_/ the hummingboard starts at pin 1 as well. What is pin 0?

**Comment by CurlyMoo** [comment]
Check this page for pin numbering in wiringX: http://wiringx.org/

So with the Hummingboard module:
`wiringX 0 == sysfs 73 == header pin 11`
Raspberry Pi:
`wiringX 0 == sysfs 17 == header pin 11`
Radxa:
`wiringX 4 == sysfs 125 == header pin 11`

Interrupts handles interrupts :smile: I think wikipedia can explain that better than me. But i'll give a shot.
https://en.wikipedia.org/wiki/Interrupt

If you want to poll changes you can do that by looping over a pin like this:

```
int a = 0, b = 0;
b = digitalRead(0);
while(1) {
   b = digitalRead(0);
   if(a != b) {
      a = b;
      printf("change\n");
   }
  usleep(1);
}
```

An interrupt does this but then at hardware level so the code becomes:

```
while(waitForInterrupt(0, 1000) > 0) {
   printf("change\n");
}
```

So as soon as the GPIO changes, the kernel will fire the wiringX `waitForInterrupt` routine making it a lot more efficient.

**Comment by garysims** [comment]
OK, so two things:

1) At the moment I have mapped the pins according to their physical pin number, I see now that you are using the GPIO number and not the physical pin number, I will change the code to do the right mapping.

2) The interrupt example doesn't work. I will need to investigate. It maybe that the sysfs stuff doesn't support this on the CI20. I am using the exact same code as for the Raspberry Pi which is all based on sysfs. The code looks the same for the Banana Pi and HummingBoard, so I don't know what is wrong, as I said I will need to investiagte some more.

However, one thing I don't get is this: if PIN 0 is being set HIGH/LOW then why do a waitForInterrupt on PIN 1???

**Comment by garysims** [comment]
So I made a link from PIN0 to PIN1 so that when the LED is HIGH a READ from PIN1 will be high.

Read.c works fine and the output from interrupt is:

interrupt
interrupt
timeout
interrupt
interrupt
timeout
interrupt
interrupt
timeout

**Comment by garysims** [comment]
Sorry, I have was having trouble with the GC... my bad, that return shouldn't be there... I will fix it...

I will alter the pins as well

File: ci20.c [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+#include <linux/spi/spidev.h>
+
+#include "wiringX.h"
+#ifndef __FreeBSD__
+	#include "i2c-dev.h"
+#endif
+#include "ci20.h"
+
+#define NUM_PINS		16
+#define PATH_MAX 64
```
Comment (CurlyMoo) [inline ci20.c]: Can you change this into this, but put in into `wiringX.h`

```
#if !defined(PATH_MAX)
    #if defined(_POSIX_PATH_MAX)
        #define PATH_MAX _POSIX_PATH_MAX
    #else
        #define PATH_MAX 1024
    #endif
#endif
```

File: ci20.c [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+	}
+
+	if(ci20ValidGPIO(pin) != 0) {
+		wiringXLog(LOG_ERR, "ci20->digitalWrite: Invalid pin number %d (0 >= pin <= 31)", pin);
+		return -1;
+	}	
+	
+	snprintf(path, PATH_MAX, "/sys/class/gpio/gpio%d/value", pinToGpio[pin]);
+	fd = open(path, O_WRONLY);
+	if (fd == -1) {
```
Comment (CurlyMoo) [inline ci20.c]: Can you make sure the coding style is consistent between all modules:
`if (` > `if(`

File: ci20.c [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+		wiringXLog(LOG_ERR, "ci20->digitalWrite: Failed to open gpio value for writing", pin);
+		return(-1);
+	}
+
+	if ( write(fd, &s_values_str[LOW == value ? 0 : 1], 1) != 1 ) {
+		wiringXLog(LOG_ERR, "ci20->digitalWrite: Failed to write value", pin);
+		return(-1);
```
Comment (CurlyMoo) [inline ci20.c]: `if ( write ... != 1 ) {` > `if(write ... != 1) {`

**Comment by CurlyMoo** [comment]
I was also wondering is there is no implementation of this module with memory mapping the GPIOs? All other wiringX module use memory mapping for digitalRead and digitalWrite, so i prefer this for all modules.

**Comment by garysims** [comment]
I looked for the memory mapping information and couldn't find any. The only published way to access the GPIO is via sysfs... If you can find any info then I am willing to take a look.

**Comment by CurlyMoo** [comment]
Is there no documentation about the SoC in regard to memory areas.

**Comment by garysims** [comment]
Not that I could find. I found the Ingenic JZ4780 datasheet but I couldn't find the info.

**Comment by CurlyMoo** [comment]
http://mipscreator.imgtec.com/CI20/hardware/soc/JZ4780_PM.pdf
Page 586

**Comment by garysims** [comment]
Yeah, I just found that... OK, I probably won't have time to look at this now for a while, don't you want to merge the code now and then maybe I can update it later with a memory mapped version?

**Comment by CurlyMoo** [comment]
Yes, i would like to, but the problem is that the current implemenation doesn't guarantee modularity between platforms. So developers might state that their program works on all wiringX platforms but because of the sysfs speed limit, it might not.

I, however, think that the memory mapping can be quite easy looking at the table on page 641 / 642.

**Comment by garysims** [comment]
Page 641 / 642?

**Comment by CurlyMoo** [comment]
Yes

**Comment by garysims** [comment]
Those pages are after the GPIO section and in the SSI section???

**Comment by CurlyMoo** [comment]
I know. If you look at the table you'll see that in order to use a pin as GPIO output you need to
1. Set mask memory bit to 1.
2. Set PAT1 bit to 0.
3. For HIGH set PAT0 bit to 1, for LOW set PAT0 bit to 0.

**Comment by garysims** [comment]
OK, the implementation using memory mapping is now ready. All tests pass.

**Comment by CurlyMoo** [comment]
Pretty damn awesome how quickly you got this working :+1: Does the interrupt example also work now?

**Comment by garysims** [comment]
Yeah, the interrupt example works. I think it did all the time, but all the low-level writing I was doing to the GPIOs got everything in a bit of a mess. After a reboot of the board it worked without any problems.

**Comment by CurlyMoo** [comment]
Thanks! Can you also update the other documentation pages and the `gh-pages` branch?

File: CMakeLists.txt [hunk 5]
```diff
@@ -25,6 +25,12 @@ if(${CMAKE_SYSTEM_PROCESSOR} MATCHES "armv7l")
 	set(CMAKE_C_FLAGS "-Ofast -fPIC -mfloat-abi=hard -mfpu=vfp -march=armv7-a -ffast-math -Wfloat-equal -Wshadow -Wcast-align -Wstrict-overflow=5 -Wwrite-strings -Waggregate-return -Wswitch-default -Wswitch-enum -Wformat=2 -g -Wall -Wno-sign-conversion -Wunreachable-code -Wstrict-prototypes")
 endif()
 
+if(${CMAKE_SYSTEM_PROCESSOR} MATCHES "^mips")
+	set(CPACK_DEBIAN_PACKAGE_ARCHITECTURE armhf)
```
Comment (ZubairLK) [inline CMakeLists.txt:29]: If(mips) set armhf?

I don't fully understand what is going on. But is that ok?

Also, 
mips generally refers to big endian. CI20 is mipsel (little endian)..

**Comment by CurlyMoo** [comment]
I'm also wondering. Is this port already being discussed somewhere on a forum for this device?

File: CMakeLists.txt (hunk 5 above)
Comment (garysims) [inline CMakeLists.txt:29]: I know less than zero about Debian packages, I am sure armhf is wrong, but I didn't know what to put there?

Does CMAKE distinguish between mips and mipsel? Can you suggest a better test for the SYSTEM PROCESSOR line?

**Comment by garysims** [comment]
Yes, as I had to get some help with the register stuff from the MIPS Creator CI20 Google Group: https://groups.google.com/forum/#!topic/mips-creator-ci20/tmmPLmvwLSY

File: CMakeLists.txt (hunk 5 above)
Comment (jcowgill) [inline CMakeLists.txt:29]: You want to set `CPACK_DEBIAN_PACKAGE_ARCHITECTURE` to `mipsel` (that's the only Debian arch which will run on the CI20 anyway). `CMAKE_SYSTEM_PROCESSOR` is set to the output of `uname -m` which doesn't distinguish between big and little endian so you can just leave the if check as it is.

Comment (garysims) [inline CMakeLists.txt:29]: uname -m reports mips... I will change the package arch now...

**Comment by CurlyMoo** [comment]
Can any of you try if pilight works on the MIPS CI20? I added the wiringX module and tested compilation with a cross-compiler.

**Pull request merged**